File: src/software/ai/hl/stp/tactic/penalty_kick/penalty_kick_tactic_test.cpp
#include "software/ai/hl/stp/tactic/penalty_kick/penalty_kick_tactic.h"

#include <gtest/gtest.h>

#include <utility>

#include "software/simulated_tests/non_terminating_validation_functions/ball_in_play_or_scored_validation.h"
#include "software/simulated_tests/non_terminating_validation_functions/ball_never_moves_backward_validation.h"
#include "software/simulated_tests/non_terminating_validation_functions/robot_not_excessively_dribbling_validation.h"
#include "software/simulated_tests/simulated_er_force_sim_play_test_fixture.h"
#include "software/simulated_tests/terminating_validation_functions/friendly_scored_validation.h"
#include "software/test_util/test_util.h"
#include "software/world/field.h"

// goalie
class PenaltyKickTacticTest : public SimulatedErForceSimPlayTestFixture,
                              public ::testing::WithParamInterface<RobotStateWithId>
{
   protected:
    TbotsProto::FieldType field_type = TbotsProto::FieldType::DIV_B;
    Field field                      = Field::createField(field_type);
    BallState ball           = BallState(field.friendlyPenaltyMark(), Vector(0, 0));
    Point initial_position   = field.friendlyPenaltyMark() + Vector(-0.1, 0);
    RobotStateWithId shooter = {
        0, RobotState(initial_position, Vector(0, 0), Angle::zero(), Angle::zero())};
    TbotsProto::AiConfig ai_config;
};

// TODO (#2232): Improve dribbling control so the ball is not lost during this test
TEST_P(PenaltyKickTacticTest, DISABLED_penalty_kick_test)
{
    RobotStateWithId enemy_robot = GetParam();

    auto tactic               = std::make_shared<PenaltyKickTactic>(ai_config);
    static RobotId shooter_id = 0;
    setTactic(shooter_id, tactic);

    std::vector<ValidationFunction> terminating_validation_functions = {
        friendlyScored,
    };

    std::vector<ValidationFunction> non_terminating_validation_functions = {
        ballInPlay,
        [](std::shared_ptr<World> world_ptr, ValidationCoroutine::push_type& yield) {
            ballNeverMovesBackward(world_ptr, yield);
        },
        [](std::shared_ptr<World> world_ptr, ValidationCoroutine::push_type& yield) {
            robotNotExcessivelyDribbling(shooter_id, world_ptr, yield);
        }};

    runTest(field_type, ball, {shooter}, {enemy_robot}, terminating_validation_functions,
            non_terminating_validation_functions, Duration::fromSeconds(10));
}

// TODO (#2519): fix and re-enable
TEST_F(PenaltyKickTacticTest, DISABLED_penalty_no_goalie)
{
    auto tactic               = std::make_shared<PenaltyKickTactic>(ai_config);
    static RobotId shooter_id = 0;
    setTactic(shooter_id, tactic);

    std::vector<ValidationFunction> terminating_validation_functions = {
        friendlyScored,
    };

    std::vector<ValidationFunction> non_terminating_validation_functions = {
        ballInPlay,
        [](std::shared_ptr<World> world_ptr, ValidationCoroutine::push_type& yield) {
            ballNeverMovesBackward(world_ptr, yield);
        },
        [](std::shared_ptr<World> world_ptr, ValidationCoroutine::push_type& yield) {
            robotNotExcessivelyDribbling(shooter_id, world_ptr, yield);
        }};

    auto enemy_robots = TestUtil::createStationaryRobotStatesWithId({Point(0, -2.5)});

    runTest(field_type, ball, {shooter}, enemy_robots, terminating_validation_functions,
            non_terminating_validation_functions, Duration::fromSeconds(10));
}

INSTANTIATE_TEST_CASE_P(
    RobotLocations, PenaltyKickTacticTest,
    ::testing::Values(
        // enemy robot stationary at centre of goal
        RobotStateWithId{0, RobotState(Field::createSSLDivisionBField().enemyGoalCenter(),
                                       Vector(0, 0), Angle::half(), Angle::zero())},

        // enemy robot stationary left of net
        RobotStateWithId{0,
                         RobotState(Field::createSSLDivisionBField().enemyGoalpostNeg(),
                                    Vector(0, 0), Angle::half(), Angle::zero())},
        // enemy robot stationary right of net
        RobotStateWithId{0,
                         RobotState(Field::createSSLDivisionBField().enemyGoalpostPos(),
                                    Vector(0, 0), Angle::half(), Angle::zero())},
        // enemy robot left of net but moving right
        RobotStateWithId{0,
                         RobotState(Field::createSSLDivisionBField().enemyGoalpostNeg(),
                                    Vector(0, 1.2), Angle::half(), Angle::zero())},
        // enemy robot right of net but moving left
        RobotStateWithId{0,
                         RobotState(Field::createSSLDivisionBField().enemyGoalpostPos(),
                                    Vector(0, -1.2), Angle::half(), Angle::zero())}));

File: src/software/ai/hl/stp/tactic/penalty_kick/penalty_kick_tactic_test.py
import pytest

import software.python_bindings as tbots_cpp
from proto.import_all_protos import PenaltyKickTactic

from software.simulated_tests.validation.friendly_team_scored import (
    FriendlyTeamEventuallyScored,
)
from software.simulated_tests.validation.excessive_dribbling import (
    NeverExcessivelyDribbles,
)
from software.simulated_tests.validation.ball_moves_in_direction import (
    BallAlwaysMovesForward,
)
from proto.message_translation.tbots_protobuf import create_world_state
from software.simulated_tests.simulated_test_fixture import (
    pytest_main,
)


@pytest.mark.parametrize(
    "enemy_robot_location, enemy_robot_velocity",
    [
        # enemy robot stationary at centre of goal
        (
            tbots_cpp.Field.createSSLDivisionBField().enemyGoalCenter(),
            tbots_cpp.Vector(0, 0),
        ),
        # enemy robot stationary left of net
        (
            tbots_cpp.Field.createSSLDivisionBField().enemyGoalpostNeg(),
            tbots_cpp.Vector(0, 0),
        ),
        # enemy robot stationary right of net
        (
            tbots_cpp.Field.createSSLDivisionBField().enemyGoalpostPos(),
            tbots_cpp.Vector(0, 0),
        ),
        # enemy robot left of net but moving right
        (
            tbots_cpp.Field.createSSLDivisionBField().enemyGoalpostNeg(),
            tbots_cpp.Vector(0, 1.2),
        ),
        # enemy robot right of net but moving left
        (
            tbots_cpp.Field.createSSLDivisionBField().enemyGoalpostPos(),
            tbots_cpp.Vector(0, -1.2),
        ),
    ],
)
@pytest.mark.skip(
    "Disabling this test because of poor dribbling controls, does not consistently score goal. TODO (#2232)"
)
def test_penalty_kick(
    enemy_robot_location, enemy_robot_velocity, simulated_test_runner
):
    field = tbots_cpp.Field.createSSLDivisionBField()
    ball_initial_pos = field.friendlyPenaltyMark()

    def setup(*args):
        shooter_position = ball_initial_pos - tbots_cpp.Vector(0.1, 0)

        simulated_test_runner.set_world_state(
            create_world_state(
                blue_robot_locations=[shooter_position],
                yellow_robot_locations=[enemy_robot_location],
                ball_location=ball_initial_pos,
                ball_velocity=tbots_cpp.Vector(0, 0),
            ),
        )

        simulated_test_runner.set_tactics(blue_tactics={0: PenaltyKickTactic()})

    eventually_validation_sequence_set = [
        [
            FriendlyTeamEventuallyScored(),
        ]
    ]

    # For RoboCup SSL rules: ball never moves backward
    always_validation_sequence_set = [
        [BallAlwaysMovesForward(ball_initial_pos), NeverExcessivelyDribbles()]
    ]

    simulated_test_runner.run_test(
        setup=setup,
        inv_eventually_validation_sequence_set=eventually_validation_sequence_set,
        inv_always_validation_sequence_set=always_validation_sequence_set,
        test_timeout_s=10,
    )


if __name__ == "__main__":
    pytest_main(__file__)
